memory_model/src/mmap.rs

```diff
@@ -291,6 +291,54 @@ impl MemoryMapping {
         // overflow. However, it is possible to alias.
         std::slice::from_raw_parts_mut(self.addr, self.size)
     }
+
+    /// Calls madvise with MADV_REMOVE to take back pages from the guest. If read from after this
+    /// call, the memory will contain only zeroes.
+    /// To learn more about madvise, read this manual page:
+    /// http://man7.org/linux/man-pages/man2/madvise.2.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use memory_model::MemoryMapping;
+    ///   let mut mem_map = MemoryMapping::new(1024).unwrap();
+    ///   assert!(mem_map.write_obj(123u32, 0).is_ok());
+    ///   assert!(mem_map.remove_range(0, 32).is_ok());
+    ///   // The kernel is now advised that the first 32 bytes of `mem_map` can be removed.
+    ///   assert_eq!(mem_map.read_obj::<u32>(32).unwrap(), 0);
+    /// ```
+    pub fn remove_range(&self, mem_offset: usize, count: usize) -> Result<()> {
+        self.range_end(mem_offset, count)
+            .map_err(|_| Error::InvalidRange(mem_offset, count))?;
+        let ret = unsafe {
+            // This is safe for the same reason that write_obj is safe:
+            // `madvise`-ing the kernel that some memory can be removed is,
+            // from the point of view of a userspace process, the same as
+            // setting the memory to 0 -- which is fine as long as we make sure
+            // that we get the bound check correct (which was done
+            // through the call to range_end previously).
+            libc::madvise(
+                (self.addr as usize + mem_offset) as *mut _,
+                count,
+                libc::MADV_REMOVE,
+            )
+        };
+        if ret < 0 {
+            Err(Error::InvalidRange(mem_offset, count))
```

Comment:
If `madvise` fails, it's not necessarily an invalid memory range. According to the [man page](http://man7.org/linux/man-pages/man2/madvise.2.html), `madvise(MADV_REMOVE)` can fail when:
```
       EACCES advice is MADV_REMOVE, but the specified address range is not
              a shared writable mapping.
       EINVAL advice is MADV_DONTNEED or MADV_REMOVE and the specified
              address range includes locked, Huge TLB pages, or VM_PFNMAP
              pages.
```
so it looks like the best course of action would be to check `errno`. You can get it with `std::io::Error::last_os_error()`. You'd also need a different error type.

Reply:
You should actually use `Error::SystemCallFailed(io::Error)`.

Reply:
Ok, sure